src/librustc/session/mod.rs

```diff
@@ -825,10 +828,17 @@ impl Session {
         }
     }
 
+    #[inline(never)]
+    #[cold]
+    fn profiler_active<F: FnOnce(&mut SelfProfiler) -> ()>(&self, f: F) {
+        let mut profiler = self.self_profiling.borrow_mut();
+        f(&mut profiler);
+    }
+
+    #[inline(always)]
     pub fn profiler<F: FnOnce(&mut SelfProfiler) -> ()>(&self, f: F) {
-        if self.opts.debugging_opts.self_profile || self.opts.debugging_opts.profile_json {
-            let mut profiler = self.self_profiling.borrow_mut();
-            f(&mut profiler);
+        if unlikely!(self.self_profiling_active) {
```

Comment:
LLVM already assigns reduced branch weights to branches postdominated by calls to cold functions, so I wouldn't expect this `unlikely!()` to have an effect here.

Reply:
I did find a case where it didn't though, so I used `likely` and `unlikely` just in case.

Reply:
I'd really like to have PGO for the Rust compiler at some point `:)`